tests/e2e/test_e2e_template.py

```diff
@@ -0,0 +1,284 @@
+    TIMEOUT_MAKE_CLEAN_NOTEBOOKS,
+    TIMEOUT_MAKE_DOWNLOAD_NOTEBOOKS,
+    TIMEOUT_MAKE_SETUP,
+    TIMEOUT_MAKE_UPLOAD_CODE,
+    TIMEOUT_MAKE_UPLOAD_DATA,
+    TIMEOUT_MAKE_UPLOAD_NOTEBOOKS,
+    TIMEOUT_NEURO_LS,
+    TIMEOUT_NEURO_PS,
+)
+
+from .conftest import (
+    DEFAULT_ERROR_PATTERNS,
+    DEFAULT_TIMEOUT_LONG,
+    DEFAULT_TIMEOUT_SHORT,
+    N_FILES,
+    cleanup_local_dirs,
+    get_logger,
+    measure_time,
+    neuro_ls,
+    neuro_ps,
+    neuro_rm_dir,
+    repeat_until_success,
+    run,
+)
+
+
+log = get_logger()
+
+
+def test_project_structure() -> None:
+    dirs = {f.name for f in Path().iterdir() if f.is_dir()}
+    assert dirs == {MK_DATA_PATH, MK_CODE_PATH, MK_NOTEBOOKS_PATH}
+    files = {f.name for f in Path().iterdir() if f.is_file()}
+    assert files == {
+        "Makefile",
+        "README.md",
+        "LICENSE",
+        PROJECT_APT_FILE_NAME,
+        PROJECT_PIP_FILE_NAME,
+        "setup.py",
+        "setup.cfg",
+        ".gitignore",
+    }
+
+
+def test_make_help_works() -> None:
+    out = run("make help", debug=True)
+    assert "setup" in out, f"not found in output: `{out}`"
+
+
+@pytest.mark.run(order=1)
+def test_make_setup() -> None:
+
+    # TODO: test also pre-installed APT packages
+    apt_deps_messages = [
+        f"Selecting previously unselected package {entry}"
+        for entry in PACKAGES_APT_CUSTOM
+    ]
+    # TODO: test also pre-installed PIP packages
+    pip_deps_entries = sorted(
+        [entry.replace("==", "-").replace("_", "-") for entry in PACKAGES_PIP_CUSTOM]
+    )
+    pip_deps_message = r"Successfully installed [^\n]* " + r"[^\n]*".join(
+        pip_deps_entries
+    )
+
+    expected_patterns = [
+        # run
+        r"Status:[^\n]+running",
+        # copy apt.txt
+        f"Copy 'file://.*{PROJECT_APT_FILE_NAME}",
+        rf"'{PROJECT_APT_FILE_NAME}' \d+B",
+        # copy pep.txt
+        f"Copy 'file://.*{PROJECT_PIP_FILE_NAME}",
+        rf"'{PROJECT_PIP_FILE_NAME}' \d+B",
+        # apt-get install
+        *apt_deps_messages,
+        r"APT requirements installation completed",
+        # pip install
+        # (pip works either with stupid progress bars, or completely silently)
+        pip_deps_message,
+        r"PIP requirements installation completed",
+        # neuro save
+        r"Saving .+ \->",
+        r"Creating image",
+        r"Image created",
+        r"Pushing image .+ => .+",
+        r"image://.*",
+        # neuro kill
+        "neuro kill",
+        r"job\-[^\n]+",
+        # success
+        r"Project setup completed",
+    ]
+
+    make_cmd = "make setup"
+    with measure_time(make_cmd):
+        run(
+            make_cmd,
+            debug=True,
+            timeout=TIMEOUT_MAKE_SETUP,
+            expect_patterns=expected_patterns,
+            # TODO: add specific error patterns
+            stop_patterns=DEFAULT_ERROR_PATTERNS,
+        )
+
+
+@pytest.mark.run(order=2)
+def test_make_upload_clean_code() -> None:
```

Comment:
this test flakes sometimes: https://circleci.com/gh/neuromation/cookiecutter-neuro-project/106

Reply:
Let's put `@flake(retries=3)` on it?

Reply:
added global option `--reruns=2`: we all know how e2e tests flake

Reply:
this means each failed test will run once again. Usually it helps